### DISCUSSION TITLE
QuantizedMeshLoader with TerrainLayer

### DISCUSSION THREAD

**Discussion by jlumsden-mts** [visgl/deck.gl#7494] [Q&A]
Hi there, we've had some good success using the terrain layer with RGB (terrarium) tiles but I've been investigating using quantized mesh tiles due to their smaller size and (hopefully) improved performance due to the tiles being pre-processed.

There seems to have been some work done on this in the past, there's a QuantizedMeshLoader in loaders.gl which I've confirmed works well but I'm having trouble getting anything to display visually. I have tested my tiles and tile server in Cesium and they look fine.

Things I have tried:
1. Adding `loaders: [QuantizedMeshLoader]` and using a ".terrain" URL for the elevation data. This doesn't produce any errors and I can debug the loader and see it is downloading and decoding the tiles fine but nothing displays
2. Trying these changes from this draft PR: https://github.com/visgl/deck.gl/pull/5061 - again nothing displays.
3. Trying this custom layer here: https://github.com/kylebarron/quantized-mesh-encoder/blob/master/site/src/quantized-mesh-layer.js - I can get this displaying if I clone this repo verbatim but the deck.gl API has changed since then and I've had trouble updating it. Even though it does display, mountains and other terrain appear to be in the wrong spot compared to the texture.

It looks like this can be done and has been done (certainly by @kylebarron) but I'm not familiar enough with deck.gl to know where I should be looking for the disconnect.

The follow up question I have on this is if there is a simple way to apply a terrain exaggeration factor similar to what you might do by adjusting the elevation decoder for the RGB terrain. It looks like this can be achieved by adjusting the Z-scaling of the model matrix but this would need to be exposed externally.

**Comment by kylebarron** [comment]
> nothing displays

you might need to check the coordinate system and orientation of your data. 



> deck.gl API has changed since then and I've had trouble updating it
> wrong spot compared to the texture

I don't think the API here has changed that much. What do you mean "wrong spot"?

**Reply by jlumsden-mts** [reply]
Thanks for responding kyle, I wasn't sure how to describe the 'wrong spot' issue as it didn't look like an obvious mesh orientation or flipping problem but looking at it further my mesh tiles were in geodetic TMS format which was the main source of my problems. The individual mesh also needed flipping in the Y-axis.

I was generating the mesh tiles using ctb-tile from https://hub.docker.com/r/tumgis/ctb-quantized-mesh

Perhaps the easier way to go would be to use some of my existing terrarium RGB tiles and convert the tiles to mesh format (using https://github.com/kylebarron/quantized-mesh-encoder) rather than going back to the source GeoTIFF. I'll give that a try as at least I know those tiles are OK and in the right spot.

**Reply by kylebarron** [reply]
> geodetic TMS format

With relatively recent changes in the TileLayer, it should be easier to subclass that layer to use a global geodetic tiling system, which I think should maybe be more accurate/faster in globe mode. I've wanted to explore that for a long time but never got a chance.

> The individual mesh also needed flipping in the Y-axis.

Yeah I never really figured out when I needed to flip the Y axis or not. 

I've personally found delatin+quantized mesh encoder to be efficient

**Comment by amuedespacher** [comment]
I reused some of the code from @kylebarron (thanks!) provided under [quantized-mesh-encoder](https://github.com/kylebarron/quantized-mesh-encoder/blob/master/site/src/quantized-mesh-layer.js), using [`tin-terrain`](https://github.com/heremaps/tin-terrain) to generate static `.terrain` files with quantized mesh. In order to make my terrain display nicely, I had to consider the following:
- [`getModelMatrix`](https://github.com/kylebarron/quantized-mesh-encoder/blob/0812c1830fd3aefe2e6ca27bf15b7bfe6cb5f8a0/site/src/quantized-mesh-layer.js#L103): Read `x`, `y`, and `z` from `tile.index` instead of `tile`.
- Define the `skirtHeight` as `quantizedMesh.skirtHeight` instead of `terrain.skirtHeight` (see [issue](https://github.com/visgl/loaders.gl/issues/3024)).
- When working with `tin-terrain`, which generates TMS-based tiles and not XYZ:
  - Reverse the `y` axis with: `Math.pow(2, z) - y - 1`
  - In the options of the `QuantizedMeshLoader`, set `quantizedMesh.bounds: [0, 1, 1, 0]` (instead of the default `[0, 0, 1, 1]`).
  - Ensure that the `ImageLoader` (loading the texture tiles) has `imagebitmap.imageOrientation` set to `flipY`.
